crates/matrix-sdk-sqlite/src/lib.rs

```diff
@@ -0,0 +1,111 @@
+// Copyright 2022 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use async_trait::async_trait;
+use deadpool_sqlite::{CreatePoolError, Object as SqliteConn};
+use matrix_sdk_crypto::{store::Result, CryptoStoreError};
+use matrix_sdk_store_encryption::StoreCipher;
+use rusqlite::OptionalExtension;
+use thiserror::Error;
+use tracing::error;
+
+#[cfg(feature = "crypto-store")]
+mod crypto_store;
+mod utils;
+
+pub use self::crypto_store::SqliteCryptoStore;
+use self::utils::SqliteObjectExt;
+
+/// All the errors that can occur when opening a sled store.
+#[derive(Error, Debug)]
+#[non_exhaustive]
+pub enum OpenStoreError {
+    /// An error occurred with the crypto store implementation.
+    #[cfg(feature = "crypto-store")]
+    #[error(transparent)]
+    Crypto(#[from] CryptoStoreError),
+
+    /// An error occurred with sqlite.
+    #[error(transparent)]
+    Sqlite(#[from] CreatePoolError),
+}
+
+async fn get_or_create_store_cipher(passphrase: &str, conn: &SqliteConn) -> Result<StoreCipher> {
```

Comment:
I think making a separate PR would be good, since this one is already pretty large.

Reply:
That's now implemented in https://github.com/matrix-org/matrix-rust-sdk/pull/1434. We'll just need to have a constructor for the store that accepts a key instead of a passphrase.